cmake/git-versioning.cmake

```
@@ -0,0 +1,30 @@
+find_program(GIT git)
+set(GIT_TAG "v0")
+set(GIT_HASH "")
+
+if(GIT)
+    execute_process(COMMAND ${GIT} describe --tags --abbrev=0
+        WORKING_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR}
+        OUTPUT_VARIABLE _GIT_TAG
+        OUTPUT_STRIP_TRAILING_WHITESPACE
+        ERROR_QUIET
+    )
+    execute_process(COMMAND ${GIT} describe --always
+        WORKING_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR}
+        OUTPUT_VARIABLE GIT_HASH
+        OUTPUT_STRIP_TRAILING_WHITESPACE
+        ERROR_QUIET
+    )
+
+    string(FIND "${_GIT_TAG}" "fatal" find_result)
+
+    if(${find_result} EQUAL -1)
+        set(GIT_TAG "${_GIT_TAG}")
+    endif()
+endif()
+
+# Output
+add_definitions(-DGIT_TAG=${GIT_TAG})
+add_definitions(-DGIT_HASH=${GIT_HASH})
+set(CPACK_PACKAGE_VERSION ${GIT_TAG})
+message(STATUS "Version ${GIT_TAG} ${GIT_HASH}")
```

We get a lot of poop in the build logs because this is a global `add_definitions'. Maybe we can start doing better with something like this suggestion and then use the `target_add_git_versioning()` function to the main driver target. 

```suggestion
set(CPACK_PACKAGE_VERSION ${GIT_TAG})
message(STATUS "Version ${GIT_TAG} ${GIT_HASH}")
function(target_add_git_versioning tgt)
    target_compile_definitions(${tgt} 
        -DGIT_TAG=${GIT_TAG}
        -DGIT_HASH=${GIT_HASH}
    )
endfunction()
```

---

If we figure it out here we can copy the pattern around.

---

Agreed on the poop. If that is also in egrabber (and maybe others), I'd prefer to create an issue here to follow up rather than work out all the details in this PR.

---

Ah wait, now I saw the sibling comment. I effectively made those changes and the poop is gone now. How can I verify that GIT_TAG and GIT_HASH are being defined/used as intended?

---

I think I verified by grepping the build directory for those two and found that the compile commands use the correct GIT_HASH for this source, while leaving the others alone to use theirs (without the redefine). So I'm going to push this.